data/abi/abi_encode_test.go

```diff
@@ -0,0 +1,1016 @@
+				require.NoError(t, err, "ufixed encode fail")
+
+				ufixedType, err := MakeUfixedType(uint16(size), uint16(precision))
+				require.NoError(t, err, "ufixed type make fail")
+
+				decodedUfixed, err := Decode(encodedUfixed, ufixedType)
+				require.NoError(t, err, "decoding ufixed should not return error")
+				require.Equal(t, valueUfixed, decodedUfixed, "decode ufixed fail to match expected value")
+			}
+		}
+	}
+
+	// decoding test for address, randomly take 1000 tests
+	// address is type alias of byte[32], we generate address value with random 256 bit big int values
+	// we make the expected address value and decode the encoding of expected, check if they match
+	upperLimit := big.NewInt(0).Lsh(big.NewInt(1), 256)
+	for i := 0; i < 1000; i++ {
+		randomAddrInt, err := rand.Int(rand.Reader, upperLimit)
+		require.NoError(t, err, "cryptographic random int init fail")
+
+		addressBytes := randomAddrInt.Bytes()
+		address := make([]byte, 32-len(addressBytes))
+		address = append(address, addressBytes...)
+
+		var addrBytes [32]byte
+		copy(addrBytes[:], address[:32])
+
+		addressValue := MakeAddress(addrBytes)
+		addrEncode, err := addressValue.Encode()
+		require.NoError(t, err, "address encode fail")
+
+		addressDecoded, err := Decode(addrEncode, MakeAddressType())
+		require.NoError(t, err, "decoding address should not return error")
+		require.Equal(t, addressValue, addressDecoded, "decode addr not match with expected")
+	}
+
+	// bool value decoding test
+	for i := 0; i < 2; i++ {
+		boolValue := MakeBool(i == 1)
+		boolEncode, err := boolValue.Encode()
+		require.NoError(t, err, "bool encode fail")
+		boolDecode, err := Decode(boolEncode, MakeBoolType())
+		require.NoError(t, err, "decoding bool should not return error")
+		require.Equal(t, boolValue, boolDecode, "decode bool not match with expected")
+	}
+
+	// byte value decoding test, iterating through 256 valid byte value
+	for i := 0; i < (1 << 8); i++ {
+		byteValue := MakeByte(byte(i))
+		byteEncode, err := byteValue.Encode()
+		require.NoError(t, err, "byte encode fail")
+		byteDecode, err := Decode(byteEncode, MakeByteType())
+		require.NoError(t, err, "decoding byte should not return error")
+		require.Equal(t, byteValue, byteDecode, "decode byte not match with expected")
+	}
+
+	// string value decoding test, test from utf string length 1 to 100
+	// randomly take 10 utf-8 strings to make ABI string values
+	// decode the encoded expected value and check if they match
+	for length := 1; length <= 100; length++ {
+		for i := 0; i < 10; i++ {
+			utf8Str := gobberish.GenerateString(length)
+			strValue := MakeString(utf8Str)
+			strEncode, err := strValue.Encode()
+			require.NoError(t, err, "string encode fail")
+			strDecode, err := Decode(strEncode, MakeStringType())
+			require.NoError(t, err, "decoding string should not return error")
+			require.Equal(t, strValue, strDecode, "encode string not match with expected")
+		}
+	}
```

Comment:
For some of the value encoding/decoding, if the value is suitable for iteration (e.g., `address`, `uint`, `ufixed`, `byte`, `bool`), I go through iteration (or randomly take some test point in the valid space) to check if my implementation is correct.